### ISSUE TITLE
Add parallel and volatile categories to PG functions 

### ISSUE THREAD

**Issue by Algunenano** [CartoDB/dataservices-api#406]
As done in other projects (for example https://github.com/CartoDB/crankshaft/issues/181) we should add these categories to help the planner make better decisions.

**Comment by pramsey** [comment]
Yes, good stuff.

**Comment by pramsey** [comment]
Also try and find some use cases where you're pretty sure they should parallel and then see what it takes to *force* them parallel. If you can jimmy with costs and get to parallel, that's "good". If nothing you can do gets things parallel, there may be a mis-marked "unsafe" function hiding in there somewhere.

**Comment by Algunenano** [comment]
Some thoughts from the WIP: 
Most, if not all, functions are technically UNSAFE and VOLATILE as they connect and write in Redis (quota management) but, on the other hand if I do something like this:
```sql
SELECT the_geom from mytable WHERE "the_geom" && cdb_geocode_street_point('One street', 'city', 'state', 'country');
```
I'd expect only one call to `cdb_geocode_street_point` to be made. It needs testing, but I think that to have that behaviour the function needs to be at least STABLE. IMMUTABLE is out of the question here (external service and all).

Also, setting this as PARALLEL SAFE could mean several simultaneous calls to the service (here/google/mapzen) which could be an issue if there is any rate limit. RESTRICTED could be OK.

cc/ @rafatower

**Comment by rafatower** [comment]
> I'd expect only one call to cdb_geocode_street_point to be made. It needs testing, but I think that to have that behaviour the function needs to be at least STABLE.

I agree

> Also, setting this as PARALLEL SAFE could mean several simultaneous calls to the service (here/google/mapzen) which could be an issue if there is any rate limit. 

Yes. It's a potential improvement but should be done with care. Some open questions: can we restrict the throughput? at what level/granularity? is the current setup (DS servers) able to cope with the load? Are the external services able too? I think these are good reasons to keep it out of scope FTM and get back to it later on.

My 2 cents

**Comment by Algunenano** [comment]
Some differences on how the planner works depending on the function volatility:

#### VOLATILE
```sql
CREATE OR REPLACE FUNCTION geofunction_volatile (searchtext text)
RETURNS Geometry AS $$
DECLARE
BEGIN
  RAISE NOTICE 'Volatile called';
  RETURN '0101000020E6100000D5E76A2BF69961C00B24287E8C094E40'::geometry;
END;
$$ LANGUAGE 'plpgsql' VOLATILE;
```
`geofunction_volatile` is called once per row in _all_month_
```sql
SELECT "the_geom"
FROM all_month
WHERE "the_geom" && geofunction_volatile('ignored');
NOTICE:  Volatile called
NOTICE:  Volatile called
NOTICE:  Volatile called
NOTICE:  Volatile called
NOTICE:  Volatile called
NOTICE:  Volatile called
NOTICE:  Volatile called
...
...
NOTICE:  Volatile called
                                               QUERY PLAN                                                
---------------------------------------------------------------------------------------------------------
 Seq Scan on all_month  (cost=0.00..3157.14 rows=1 width=32) (actual time=0.148..185.996 rows=1 loops=1)
   Filter: (the_geom && geofunction_volatile('ignored'::text))
   Rows Removed by Filter: 10670
 Planning time: 0.108 ms
 Execution time: 186.033 ms
(5 rows)
```


#### STABLE
```sql
CREATE OR REPLACE FUNCTION geofunction_stable (searchtext text)
RETURNS Geometry AS $$
DECLARE
BEGIN
  RAISE NOTICE 'Stable called';
  RETURN '0101000020E6100000D5E76A2BF69961C00B24287E8C094E40'::geometry;
END;
$$ LANGUAGE 'plpgsql' STABLE;
```
`geofunction_stable` is called once before the index scan (so after the plan is done)
```sql
SELECT "the_geom"
FROM all_month
WHERE "the_geom" && geofunction_stable('ignored');
NOTICE:  Stable called
                                                            QUERY PLAN                                                             
-----------------------------------------------------------------------------------------------------------------------------------
 Index Scan using all_month_the_geom_idx on all_month  (cost=0.40..8.42 rows=1 width=32) (actual time=0.232..0.233 rows=1 loops=1)
   Index Cond: (the_geom && geofunction_stable('ignored'::text))
 Planning time: 0.096 ms
 Execution time: 0.270 ms
(4 rows)
```

#### IMMUTABLE
```sql
CREATE OR REPLACE FUNCTION geofunction_immutable (searchtext text)
RETURNS Geometry AS $$
DECLARE
BEGIN
  RAISE NOTICE 'Immutable called';
  RETURN '0101000020E6100000D5E76A2BF69961C00B24287E8C094E40'::geometry;
END;
$$ LANGUAGE 'plpgsql' IMMUTABLE;
```
`geofunction_immutable` is called once, pre-evaluated by the planner, then used in the index
```sql
SELECT "the_geom"
FROM all_month
WHERE "the_geom" && geofunction_immutable('ignored');
NOTICE:  Immutable called
                                                            QUERY PLAN                                                             
-----------------------------------------------------------------------------------------------------------------------------------
 Index Scan using all_month_the_geom_idx on all_month  (cost=0.15..8.17 rows=1 width=32) (actual time=0.092..0.093 rows=1 loops=1)
   Index Cond: (the_geom && '0101000020E6100000D5E76A2BF69961C00B24287E8C094E40'::geometry)
 Planning time: 0.234 ms
 Execution time: 0.137 ms
(4 rows
```

The main difference between STABLE and IMMUTABLE here is that if the plan with the IMMUTABLE function is cached then the function won't be called again. This has a positive side (avoid unneeded calls) and a negative one (if a request fails [out of quota, service is down] or it's updated a bad request might be cached.

For example:
```sql
CREATE OR REPLACE FUNCTION geofunction_immutable_all_month()
RETURNS TABLE (out_geom Geometry) AS $$
DECLARE
BEGIN
  RETURN QUERY SELECT "the_geom"
  FROM all_month
  WHERE "the_geom" && geofunction_immutable('ignored');
END;
$$ LANGUAGE 'plpgsql' IMMUTABLE;
```

The first execution calls the immutable function, but any subsequent call get optimized away:
```sql
cartodb_dev_user_3e4a6fc6-4137-4c59-bc63-066f80efb90e_db=# Select geofunction_immutable_all_month();
NOTICE:  Immutable called
          geofunction_immutable_all_month           
----------------------------------------------------
 0101000020E6100000D5E76A2BF69961C00B24287E8C094E40
(1 row)

cartodb_dev_user_3e4a6fc6-4137-4c59-bc63-066f80efb90e_db=# Select geofunction_immutable_all_month();
          geofunction_immutable_all_month           
----------------------------------------------------
 0101000020E6100000D5E76A2BF69961C00B24287E8C094E40
(1 row)
```

This doesn't happen with a STABLE function:
```sql
CREATE OR REPLACE FUNCTION geofunction_stable_all_month()
RETURNS TABLE (out_geom Geometry) AS $$
DECLARE
BEGIN
  RETURN QUERY SELECT "the_geom"
  FROM all_month
  WHERE "the_geom" && geofunction_stable('ignored');
END;
$$ LANGUAGE 'plpgsql' STABLE;
```
```sql
cartodb_dev_user_3e4a6fc6-4137-4c59-bc63-066f80efb90e_db=# Select geofunction_stable_all_month();
NOTICE:  Stable called
            geofunction_stable_all_month            
----------------------------------------------------
 0101000020E6100000D5E76A2BF69961C00B24287E8C094E40
(1 row)

cartodb_dev_user_3e4a6fc6-4137-4c59-bc63-066f80efb90e_db=# Select geofunction_stable_all_month();
NOTICE:  Stable called
            geofunction_stable_all_month            
----------------------------------------------------
 0101000020E6100000D5E76A2BF69961C00B24287E8C094E40
(1 row)

cartodb_dev_user_3e4a6fc6-4137-4c59-bc63-066f80efb90e_db=# Select geofunction_stable_all_month();
NOTICE:  Stable called
            geofunction_stable_all_month            
----------------------------------------------------
 0101000020E6100000D5E76A2BF69961C00B24287E8C094E40
(1 row)
```

Keep in mind that this works because `geofunction_immutable` in declared IMMUTABLE, not because the volatility of the function that use it; it'd also get cached if `geofunction_immutable_all_month` was declared VOLATILE.
Some extra information can be found in [plpgsql wiki page](https://www.postgresql.org/docs/current/static/plpgsql-implementation.html).

For now, as invalidating a plan in production would be a huge pain in the ass, I'd set the plpgsql functions as STABLE (plpython does not appear to cache plans so it's not as important there).

**Comment by rafatower** [comment]
### Acceptance criteria
- [ ] When building against PG10 functions in client and server are qualified with `PARALLEL<SAFE/RESTRICTED/UNSAFE>` (and volatility also).
- [ ] Extensions tests pass against PG10
- [ ] When building against PG9.5 the parallel tags are removed so that the client and server extensions can be installed and upgraded without trouble.
- [ ] Extensions tests pass against PG9.5
- [ ] Integration tests pass (if possible at all, in staging)

**Comment by rafatower** [comment]
Additional PR: https://github.com/CartoDB/dataservices-api/pull/411

**Comment by rafatower** [comment]
Integration tests against staging :v: 
```
$ python run_tests.py --host=carto-staging.com --schema=https ethervoid-rui01-owner [API_KEY]
..........................................................................
----------------------------------------------------------------------
Ran 74 tests in 42.221s

OK
```

**Comment by rafatower** [comment]
Integration tests ran against production as well.
